src/array_slice.cpp

```diff
@@ -523,20 +523,12 @@ bool increment(size_t *n, size_t *nMax, int rank) {
   return true;
 }
 
-// data_size = 1,2 for real,complex-valued array
-realnum *collapse_array(realnum *array, int *rank, size_t dims[3], direction dirs[3], volume where,
-                        int data_size = 1) {
-
-  /*--------------------------------------------------------------*/
-  /*- detect empty dimensions and compute rank and strides for    */
-  /*- collapsed array                                             */
-  /*--------------------------------------------------------------*/
-  int full_rank = *rank;
-  if (full_rank == 0) return array;
+// get the dimensions of reduced arrays
+size_t *reduce_array_dimensions(volume where, int full_rank, int &reduced_rank, size_t &reduced_grid_size, size_t dims[3], size_t stride[3], size_t reduced_stride[3], direction dirs[3], direction reduced_dirs[3]) {
```

Comment:
Probably better to make `size_t reduced_dims[3]` one of the arguments

By returning a `new` array you probably have a memory leak.

Reply:
Maybe put all of the input arguments first, and then all of the output arguments for clarity?